silver/models/documents/base.py

```diff
@@ -419,38 +422,63 @@ def get_template(self, state=None):
 
         return select_template(templates)
 
-    def generate_pdf(self, state=None):
+    def get_pdf_filename(self):
+        return '{doc_type}_{series}-{number}.pdf'.format(
+            doc_type=self.__class__.__name__,
+            series=self.series,
+            number=self.number
+        )
+
+    def get_pdf_upload_path(self):
+        path_template = getattr(
+            settings, 'SILVER_DOCUMENT_UPLOAD_PATH',
+            'documents/{provider}/{doc.kind}/{issue_date}/{filename}'
+        )
+
+        context = {
+            'doc': self,
+            'filename': self.get_pdf_filename(),
+            'provider': self.provider.slug,
+            'customer': self.customer.slug,
+            'issue_date': self.issue_date.strftime('%Y/%m/%d')
+        }
+
+        return path_template.format(**context)
+
+    def generate_pdf(self, state=None, upload=True):
```

Comment:
I would make `PDF` a `OneToOne` field and move the entire pdf generation logic to pdf model.
`get_pdf_filename`, `get_pdf_upload_path` and `generate_pdf`